File: wasm/game.go
package main

import (
	"fmt"
	"math"
	"slices"
	"strings"
	"time"

	. "libble/shared"

	"github.com/sahilm/fuzzy"
	dom "honnef.co/go/js/dom/v2"
)

func initGame() {
	debugPrint("Starting game...")

	sendToStart := func() {
		removeData(saveKeyID)
		handlePage()
	}

	// Get libble ID
	libbleIDStr := loadLibbleID()

	var app App

	// Try loading Player from LocalStorage
	if err := loadJson(saveKeyPlayer, &app.User); err != nil {
		debugPrint("Player not in LocalStorage, fetching from server...")
		type PlayerResponse struct {
			Player User `json:"player"`
		}
		var resp PlayerResponse
		if err := fetch("/game/player/"+libbleIDStr, &resp); err != nil {
			log(err, "Failed to fetch player from server")
			sendToStart()
			return
		}
		app.User = resp.Player
		debugPrint("Loaded player from server")
	} else {
		debugPrint("Loaded player from LocalStorage")
	}

	// Try loading Books from LocalStorage
	if err := loadJson(saveKeyBooks, &app.Books); err != nil || len(app.Books) == 0 {
		debugPrint("Books not in LocalStorage, fetching from server...")
		type BooksResponse struct {
			Books []UserBook `json:"books"`
		}
		var resp BooksResponse
		if err := fetch("/game/user-books/"+libbleIDStr, &resp); err != nil {
			log(err, "Failed to fetch books from server")
			sendToStart()
			return
		}
		app.Books = resp.Books
		debugPrint("Loaded %d books from server", len(app.Books))
	} else {
		debugPrint("Loaded %d books from LocalStorage", len(app.Books))
	}

	dailyQuoteID := QuoteID(NilID)
	// Try loading Quotes from LocalStorage
	if err := loadJson(saveKeyQuotes, &app.Quotes); err != nil || len(app.Quotes) <= 0 {
		fetchedQuote, err := fetchDailyQuote(&app)
		if err != nil {
			log(err, "Failed to fetch daily quote")
			sendToStart()
			return
		}

		dailyQuoteID = fetchedQuote
		app.NeedsServer = true
		debugPrint("Loaded today's quote from server")
	} else {
		debugPrint("Loaded %d quotes from LocalStorage", len(app.Quotes))
	}
	app.PopulateLookups()

	if dailyQuoteID == NilID {
		debugPrint("Picking Daily Quote")
		quoteID, err := app.PickDailyQuote()
		log(err, "Error when picking daily quote")
		if quoteID == NilID {
			sendToStart()
			return
		}
		dailyQuoteID = quoteID
	}

	debugPrint("Daily quote is %d", dailyQuoteID)

	if _, err := app.User.InitTodaysGame(app, dailyQuoteID); err != nil {
		log(err, "Failed initializing today's game")
		sendToStart()
		return
	}

	debugPrint("Today's book is %s", app.User.TodaysGame().Book.Book.CleanTitle())

	debugPrint("Setting update autocomplete")

	// convert book map to slice
	bookCount := len(app.Books)
	allBooks := make([]Book, 0, bookCount)
	for _, book := range app.Books {
		if !app.User.Settings.ScrapeOptions.ShouldScrapeQuotes(book) {
			continue
		}
		allBooks = append(allBooks, book.Book)
	}
	debugPrint("Using %d/%d books (from scrape options)", len(allBooks), len(app.Books))

	setupHTML(&app, allBooks)
}

func pickDailyQuote(data *App) (QuoteID, error) {
	if !data.NeedsServer {
		return data.PickDailyQuote()
	}
	return fetchDailyQuote(data)
}

func fetchDailyQuote(app *App) (QuoteID, error) {
	type DailyGameResponse struct {
		Quote Quote    `json:"quote"`
		Book  UserBook `json:"book"`
	}
	var dailyResp DailyGameResponse
	var dailyQuote QuoteID
	libbleID := loadLibbleID()
	if err := fetch("/game/daily/"+libbleID, &dailyResp); err != nil {
		return dailyQuote, err
	}
	app.AddQuote(dailyResp.Quote)
	return dailyResp.Quote.ID, nil
}

func setupHTML(app *App, allBooks Books) {
	doc := dom.GetWindow().Document()

	game := app.User.TodaysGame()

	defer func() {
		if r := recover(); r != nil {
			logErr("Recovered from panic setting up html:\n%v", r)
		}
	}()

	quoteElement := getElemByID(doc, "quote")
	if quoteElement != nil {
		quoteElement.SetTextContent(game.Quote.Text)
	}

	input := getElemByIDAs[*dom.HTMLInputElement](doc, "title")
	suggestions := getElemByIDAs[dom.HTMLElement](doc, "titleSuggestions")
	guessForm := getElemByID(doc, "guessForm")

	feedback := getElemByIDAs[dom.HTMLElement](doc, "feedbackBox")
	setFeedback := func(msg string, status string) {
		if feedback != nil {
			setFeedbackElem(feedback, msg, status)
		}
	}

	gameInputs := doc.GetElementsByClassName("game-input")
	submitBtn := getElemByIDAs[*dom.HTMLButtonElement](doc, "submitBtn")
	shareBtn := getElemByIDAs[*dom.HTMLButtonElement](doc, "shareBtn")
	gameProgress := getElemByIDAs[dom.HTMLElement](doc, "gameProgress")

	hintSection := getElemByIDAs[dom.HTMLElement](doc, "hintSection")
	skipBtn := getElemByIDAs[*dom.HTMLButtonElement](doc, "skipBtn")

	hints := createHints(hintSection)
	updateGameProgress := func() {
		if gameProgress == nil {
			return
		}
		var sb strings.Builder
		addEmojiStrip(game, &sb)
		gameProgress.SetTextContent(sb.String())
		// TODO: show emoji strip above book title if it starts wrapping
	}

	updateInputStates := func() {
		defer func() {
			if r := recover(); r != nil {
				logErr("Failed to update inputs")
			}
		}()
		completed := game.Completed()
		disabled := completed
		for _, e := range gameInputs {
			e.Underlying().Set("disabled", disabled)
		}

		setVisible(hintSection, !completed)

		started := game.Started()
		setVisible(skipBtn, !completed && !started)
		setEnabled(skipBtn, !completed && !started)
		for _, hint := range hints {
			if game.UsedHint(hint.kind) {
				hint.elem.Class().Add(usedHintClass)
			}
			show := !completed && started && hint.canUse(*game)
			setVisible(hint.elem, show)
			setEnabled(hint.elem, show)
		}

		updateGameProgress()

		setVisible(submitBtn, !completed)
		setVisible(shareBtn, completed)
	}

	handleRevist := func() bool {
		if !game.Completed() {
			setVisible(submitBtn, true)
			setVisible(hintSection, true)
			return true
		}
		if game.Won() {
			setFeedback("Congrats! You've already won for today,\ncome back tomorrow to play again.", FBStatusSuccess)
		} else {
			setFeedback("Looks like you didn't get it this time :(\nCome back tomorrow and try again!", FBStatusError)
		}
		input.SetPlaceholder(game.Book.Book.CleanTitle())

		setVisible(submitBtn, false)
		setVisible(hintSection, false)
		setVisible(shareBtn, true)
		shareBtn.SetDisabled(false)
		return false
	}

	handleRevist()
	updateInputStates()

	// setup submit
	guessForm.AddEventListener("submit", false, func(e dom.Event) {
		e.PreventDefault()
		if handleRevist() {
			go func() {
				onSubmit(input, app, setFeedback)
				updateInputStates()
			}()
		}
	})

	// setup skip button
	skipBtn.AddEventListener("click", false, func(e dom.Event) {
		e.PreventDefault()
		if game.Attempts() <= 0 && handleRevist() {
			// Run in goroutine to avoid blocking the event loop
			go func() {
				err := onSkip(app, setFeedback)
				log(err, "Failed skipping current quote")

				quoteElement.SetTextContent(game.Quote.Text)
				updateInputStates()
			}()
		}
	})

	// setup hints
	for _, hint := range hints {
		if hint.elem == nil {
			continue
		}
		hint.elem.AddEventListener("click", false, func(e dom.Event) {
			debugPrint("Used hint: " + string(hint.kind))
			e.PreventDefault()
			if handleRevist() {
				go func() {
					msg := hint.use(game)
					if msg != "" {
						setFeedback(msg, FBStatusHint)
					}
					updateInputStates()
					saveErr := saveNonStaticData(*app)
					log(saveErr, "Failed saving data after using hint")
				}()
			}

		})
	}

	// setup share button
	shareBtn.AddEventListener("click", false, func(e dom.Event) {
		e.PreventDefault()
		if game.Completed() {
			results := generateResultsString(game)
			err := shareText(results)
			if err == nil {
				return
			}

			err = copyToClipboard(results)
			prevText := shareBtn.TextContent()
			if err == nil {
				shareBtn.SetTextContent("📋 Copied!")
			} else {
				log(err, "Failed to share")
				shareBtn.SetTextContent("Sorry, failed to copy")
			}
			shareBtn.SetDisabled(true)

			go func() {
				time.Sleep(time.Second * 3)
				shareBtn.SetTextContent(prevText)
				shareBtn.SetDisabled(false)
			}()
		} else {
			logErr("How did you click this?")
		}
	})

	setupAutocomplete(input, suggestions, allBooks)

	// Setup user icon
	setupUserIcon()
}

const (
	// Feedback statuses
	FBStatusSuccess = "success"
	FBStatusError   = "error"
	FBStatusWarn    = "warning"
	FBStatusHint    = "hint"
)

func setFeedbackElem(e dom.HTMLElement, message string, status string) {
	if e == nil {
		return
	}

	e.SetTextContent(message)
	e.Class().SetString("feedback " + status)
}

func onSubmit(
	input *dom.HTMLInputElement,
	app *App,
	setFeedback func(msg string, status string),
) bool {

	query := strings.ToLower(strings.TrimSpace(input.Value()))
	game := app.User.TodaysGame()
	target := strings.ToLower(game.Book.Book.CleanTitle())

	defer saveNonStaticData(*app)

	if query == target {
		game.Guesses = append(game.Guesses, game.Quote.BookID)
		attempts := len(game.Guesses)
		s := ""
		if attempts > 1 {
			s = "s"
		}
		message := fmt.Sprintf("Correct! You got it in %d attempt%s", attempts, s)
		setFeedback(message, FBStatusSuccess)
		return true
	} else if bookID := app.FindBookID(query); bookID != NilID {
		if slices.Contains(game.Guesses, bookID) {
			setFeedback("You already tried that guess!", FBStatusWarn)
		} else {
			game.Guesses = append(game.Guesses, bookID)

			if len(game.Guesses) >= game.Settings.MaxGuesses {
				msg := fmt.Sprintf("Failed! The answer was \"%s\"", game.Book.Book.CleanTitle())
				setFeedback(msg, FBStatusError)
				return true
			} else {
				msg := fmt.Sprintf("Nope! Try again (%d attempts remaining)",
					game.Settings.MaxGuesses-len(game.Guesses))
				setFeedback(msg, FBStatusError)
			}
		}
	} else {
		setFeedback("That book is not in your library!", FBStatusWarn)
	}
	return false
}

func onSkip(
	app *App,
	setFeedback func(msg string, status string),
) error {
	game := app.User.TodaysGame()

	if game.Started() {
		return fmt.Errorf("Trying to skip after the game already started")
	}

	// Mark quote as seen so it won't appear again
	if !slices.Contains(app.User.SeenQuotes, game.QuoteID) {
		app.User.SeenQuotes = append(app.User.SeenQuotes, game.QuoteID)
		debugPrint("Skipping quote %d", game.QuoteID)
		if app.NeedsServer {
			syncPlayer(app.User) // Need to sync since pickDailyQuote will fetch data from server
		}
	}

	msg := fmt.Sprintf("Skipped! The answer was \"%s\"", game.Book.Book.CleanTitle())
	setFeedback(msg, FBStatusError)

	dailyQuoteID, err := pickDailyQuote(app)
	if dailyQuoteID == NilID {
		return fmt.Errorf("Failed to repick daily quote when skipping:\n%v", err)
	}
	log(err, "Issue when to repicking daily quote when skipping")
	game.QuoteID = dailyQuoteID
	game.Date = time.Now()
	saveNonStaticData(*app)
	if err := game.Init(*app); err != nil {
		return err
	}
	debugPrint("The book is now %s", game.Book.Book.CleanTitle())
	return nil
}

func addEmojiStrip(game *Game, sb *strings.Builder) {
	hintIndex := 0
	for i := 0; i < game.Settings.MaxGuesses; i++ {
		for hintIndex < len(game.Hints) {
			hint := game.Hints[hintIndex]
			if hint.GuessIndex != i {
				break
			}

			sb.WriteString(hintEmoji(hint.Kind))
			sb.WriteRune(' ')
			hintIndex++
		}

		if i < len(game.Guesses) {
			if i == len(game.Guesses)-1 && game.Won() {
				sb.WriteRune('🟩')
			} else {
				sb.WriteRune('🟥')
			}
		} else {
			sb.WriteRune('⬜')
		}
		if i < game.Settings.MaxGuesses-1 {
			sb.WriteRune(' ')
		}
	}
}

func generateResultsString(game *Game) string {
	// Create shareable text
	var shareText strings.Builder
	shareText.WriteString("📖 Libble - ")

	// Format date
	dateStr := game.Date.Format("Jan 2 2006")
	shareText.WriteString(dateStr)

	shareText.WriteRune('\n')
	addEmojiStrip(game, &shareText)

	shareText.WriteString("\n\n")
	shareText.WriteString(game.Quote.Text)

	// Copy to clipboard
	return shareText.String()
}

// TODO: Remove the end (...) on quotes
// TODO: Also search by author
// TODO: highlight indexes that matched the text
func setupAutocomplete(
	input *dom.HTMLInputElement,
	suggestionsParent dom.HTMLElement,
	allBooks Books /* available books */) {

	doc := dom.GetWindow().Document()

	type Suggestion struct {
		bookIndex  int
		titleMatch fuzzy.Match
	}

	suggestions := make([]Suggestion, 0, len(allBooks))
	currentSelection := 0
	const maxVisibleSuggestions = 8

	getBookTitle := func(suggestionIndex int) string {
		if suggestionIndex >= len(suggestions) {
			logErr("Trying to get suggestion index %d but only have %d",
				suggestionIndex, len(suggestions))
			return ""
		}
		suggestion := suggestions[suggestionIndex]
		if suggestion.bookIndex >= len(allBooks) {
			logErr("Book index %d is out of bounds of %d. How???", suggestion.bookIndex, len(allBooks))
			return ""
		}
		return allBooks[suggestion.bookIndex].CleanTitle()
	}

	updateSuggestions := func() {}

	resetSuggestions := func() {
		currentSelection = 0
		suggestions = suggestions[:0]
		updateSuggestions()
	}

	useSelection := func() {
		input.SetValue(getBookTitle(currentSelection))
		resetSuggestions()
	}

	setSelection := func(selection int) {
		currentSelection = selection
		updateSuggestions()
		input.SetValue(getBookTitle(currentSelection))
	}

	updateSuggestions = func() {
		suggestionsParent.SetInnerHTML("")

		setDisplay := func(val string) {
			suggestionsParent.Style().SetProperty("display", val, "important")
		}

		if len(suggestions) == 0 {
			setDisplay("none")
			return
		}

		for i, suggestion := range suggestions {
			li := doc.CreateElement("li")

			book := allBooks[suggestion.bookIndex]

			li.SetTextContent(book.CleanTitle())
			if i == currentSelection {
				li.Class().Add("selected")
			}
			li.AddEventListener("click", false, func(e dom.Event) {
				setSelection(i)
				useSelection()
			})
			suggestionsParent.AppendChild(li)
			if i >= maxVisibleSuggestions {
				break
			}
		}
		setDisplay("block")
	}

	input.AddEventListener("input", false, func(e dom.Event) {
		query := strings.TrimSpace(input.Value())

		matches := fuzzy.FindFrom(query, allBooks)
		count := min(len(matches), int(80))
		suggestions = suggestions[:0]
		currentSelection = 0
		for i := range count {
			match := matches[i]
			suggestions = append(suggestions, Suggestion{
				bookIndex:  match.Index,
				titleMatch: match,
			})
		}

		// Find matching books
		updateSuggestions()
	})

	input.AddEventListener("keydown", false, func(e dom.Event) {
		if suggestionsParent.InnerHTML() == "" {
			return
		}
		keyEvent := e.(*dom.KeyboardEvent)
		key := keyEvent.Key()
		// TODO: figure out how to make it select the entire text on single click

		switch key {
		// TODO: figure out how to "scroll" the suggestions so it moves the selection
		// if it goes out of being visible
		case "ArrowDown":
			e.PreventDefault()
			setSelection((currentSelection + 1) % len(suggestions))
		case "ArrowUp":
			e.PreventDefault()
			if currentSelection == 0 {
				setSelection(maxVisibleSuggestions - 1)
			} else {
				setSelection(currentSelection - 1)
			}
		case "Enter":
			e.PreventDefault()
			useSelection()
		case "Tab":
			e.PreventDefault()
			input.SetValue(getBookTitle(currentSelection))
		case "Escape":
			resetSuggestions()
		}
	})

	// Hide suggestions when clicking outside
	doc.AddEventListener("click", false, func(e dom.Event) {
		target := e.Target()
		if target != input && target != suggestionsParent {
			resetSuggestions()
		}
	})
}

// func fuzzyScore(query, ) {
// 	if (text.startsWith(query)) {
// 		return 1.0;
// 	}
//
// 	// Check if any word in text starts with query
// 	const words = text.split(/\s+/);
// 	for (let word of words) {
// 		if (word.startsWith(query)) {
// 			return 0.9;
// 		}
// 	}
//
// 	// Check if query is contained in text
// 	if (text.includes(query)) {
// 		return 0.8;
// 	}
// }

func LevenshteinDistance(s, t string) int {
	r1, r2 := []rune(s), []rune(t)
	column := make([]int, 1, 64)

	for y := 1; y <= len(r1); y++ {
		column = append(column, y)
	}

	for x := 1; x <= len(r2); x++ {
		column[0] = x

		for y, lastDiag := 1, x-1; y <= len(r1); y++ {
			oldDiag := column[y]
			cost := 0
			if r1[y-1] != r2[x-1] {
				cost = 1
			}
			column[y] = min(column[y]+1, column[y-1]+1, lastDiag+cost)
			lastDiag = oldDiag
		}
	}
	return column[len(r1)]
}

func LevenshteinDistanceNorm(s1, s2 string) float64 {
	distance := LevenshteinDistance(s1, s2)
	maxLength := math.Max(float64(len(s1)), float64(len(s2)))

	if maxLength == 0 { // Handle case where both strings are empty
		return 0.0
	}
	return float64(distance) / maxLength
}

type Books []Book

func (b Books) String(i int) string {
	if i >= 0 && i < len(b) {
		return b[i].CleanTitle()
	}
	logErr("Fuzzy search is trying to use index %d", i)
	return ""
}

func (b Books) Len() int {
	return len(b)
}

// func findMatchingBooks(query string, book []Book) []Suggestion {
// 	if query == "" {
// 		return nil
// 	}
//
// 	fuzzy.FindFromNoSort(query, data.Books)
// 	// matches := make([]Match, 0, len(data.Books))
// 	//
// 	// for _, userBook := range data.Books {
// 	// 	book := userBook.Book
// 	//
// 	// 	title := book.CleanTitle()
// 	// 	if strings.ToLower(query) != query { // only ignore case if the user used uppercase
// 	// 		title = strings.ToLower(title)
// 	// 	}
// 	//
// 	// 	score := LevenshteinDistance(query, title)
// 	// 	logg.Debugf("%s has score %d with %s\n", title, score, query)
// 	// 	if score > 0 {
// 	// 		matches = append(matches, Match{
// 	// 			book:  book,
// 	// 			score: score,
// 	// 		})
// 	// 	}
// 	// }
// 	// slices.SortFunc(matches, func(a Match, b Match) int {
// 	// 	return b.score - a.score
// 	// })
//
// 	return matches
// }

File: site/js/game.js
let books = [];
let allQuotes = [];
let currentQuote = null;
let usedQuoteIds = new Set();
let attemptCount = 0;
const MAX_ATTEMPTS = 5;
let previousGuesses = [];
let hintsUsed = [];

// Load data from JSON files
async function loadData() {
    try {
        const [booksResponse, quotesResponse] = await Promise.all([
            fetch("./books.json"),
            fetch("./quotes.json"),
        ]);

        const booksData = await booksResponse.json();
        const quotesData = await quotesResponse.json();

        // Process books - only include books with valid dates_read
        books = booksData
            .filter((book) =>
                book.dates_read && book.dates_read.length > 0 &&
                book.dates_read[0] !== "not set"
            )
            .map((book) => {
                // Get the most recent read date
                const mostRecentDate = book.dates_read[
                    book.dates_read.length - 1
                ];
                return {
                    book_id: book.book_id,
                    title: book.raw_title.split("\n")[0]
                        .trim(),
                    author: book.author.split(",").reverse()
                        .join(" ").trim(),
                    date_read: mostRecentDate,
                };
            });

        // Flatten quotes array and match with books
        allQuotes = quotesData.flat().map((quote) => {
            const book = books.find((b) => b.book_id === quote.book_id);
            return {
                quote_id: quote.quote_id,
                text: quote.text,
                book_id: quote.book_id,
                book: book,
            };
        }).filter((q) => q.book); // Only keep quotes with matching books

        console.log(
            `Loaded ${books.length} books and ${allQuotes.length} quotes`,
        );

        // Load used quotes from localStorage
        const storedUsed = localStorage.getItem("usedQuoteIds");
        if (storedUsed) {
            usedQuoteIds = new Set(JSON.parse(storedUsed));
        }

        // Reset if all quotes have been used
        if (usedQuoteIds.size >= allQuotes.length) {
            usedQuoteIds.clear();
            localStorage.removeItem("usedQuoteIds");
        }

        // Select today's quote
        selectDailyQuote();
    } catch (error) {
        console.error("Error loading data:", error);
        // Fallback data
        books = [
            { title: "Pride and Prejudice", author: "Jane Austen" },
            { title: "1984", author: "George Orwell" },
        ];
        currentQuote = {
            text: "It is a truth universally acknowledged...",
            book: books[0],
        };
        document.getElementById("quote").textContent = currentQuote.text;
    }
}

function selectDailyQuote() {
    // Get today's date as a seed
    const today = new Date();
    const daysSinceEpoch = Math.floor(
        today.getTime() / (1000 * 60 * 60 * 24),
    );

    // Get available quotes (not used yet)
    const availableQuotes = allQuotes.filter((q) =>
        !usedQuoteIds.has(q.quote_id)
    );

    if (availableQuotes.length === 0) {
        // All quotes used, reset
        usedQuoteIds.clear();
        localStorage.removeItem("usedQuoteIds");
        return selectDailyQuote();
    }

    // Use day as seed for consistent daily selection
    const index = daysSinceEpoch % availableQuotes.length;
    currentQuote = availableQuotes[index];

    // Mark as used
    usedQuoteIds.add(currentQuote.quote_id);
    localStorage.setItem("usedQuoteIds", JSON.stringify([...usedQuoteIds]));

    // Reset attempt count for new quote
    attemptCount = 0;
    previousGuesses = [];
    hintsUsed = [];

    // Reset hint buttons
    document.getElementById("timeHintBtn").disabled = false;
    document.getElementById("hintDisplay").textContent = "";
    document.getElementById("hintDisplay").style.display = "none";

    // Display the quote
    document.getElementById("quote").textContent = currentQuote.text;
}

document.getElementById("guessForm").addEventListener("submit", function (e) {
    e.preventDefault();

    if (!currentQuote) return;

    const title = document.getElementById("title").value.trim()
        .toLowerCase();
    const author = document.getElementById("author").value.trim()
        .toLowerCase();
    const feedbackBox = document.getElementById("feedbackBox");

    const correctTitle = currentQuote.book.title.toLowerCase();
    const correctAuthor = currentQuote.book.author.toLowerCase();

    // Check if the book exists in the library
    const isValidBook = books.some((book) =>
        book.title.toLowerCase() === title &&
        book.author.toLowerCase() === author
    );

    if (!isValidBook) {
        feedbackBox.textContent = `⚠️ That book is not in your library!`;
        feedbackBox.className = "feedback warning";
        return;
    }

    // Check if this guess was already made
    const guessKey = `${title}|${author}`;
    if (previousGuesses.includes(guessKey)) {
        feedbackBox.textContent = `⚠️ You already tried that guess!`;
        feedbackBox.className = "feedback warning";
        return;
    }

    previousGuesses.push(guessKey);
    attemptCount++;

    if (title === correctTitle && author === correctAuthor) {
        feedbackBox.textContent =
            `🎉 Correct! You got it in ${attemptCount} attempt${
                attemptCount === 1 ? "" : "s"
            }!`;
        feedbackBox.className = "feedback success";
        disableInputs();
    } else if (attemptCount >= MAX_ATTEMPTS) {
        feedbackBox.textContent =
            `❌ Failed! The answer was "${currentQuote.book.title}" by ${currentQuote.book.author}`;
        feedbackBox.className = "feedback error";
        disableInputs();
    } else if (author === correctAuthor && title !== correctTitle) {
        feedbackBox.textContent =
            `✍️ You got the author! Now guess the title. (${attemptCount}/${MAX_ATTEMPTS})`;
        feedbackBox.className = "feedback partial";
    } else if (title === correctTitle && author !== correctAuthor) {
        feedbackBox.textContent =
            `📚 You got the title! Now guess the author. (${attemptCount}/${MAX_ATTEMPTS})`;
        feedbackBox.className = "feedback partial";
    } else {
        feedbackBox.textContent =
            `❌ Nope! Try again. (${attemptCount}/${MAX_ATTEMPTS})`;
        feedbackBox.className = "feedback error";
    }
});

function disableInputs() {
    document.getElementById("title").disabled = true;
    document.getElementById("author").disabled = true;
    document.querySelector(".submit-btn").disabled = true;
    document.getElementById("timeHintBtn").disabled = true;
}

// Time hint functionality
document.getElementById("timeHintBtn").addEventListener("click", function () {
    if (!currentQuote || hintsUsed.includes("time")) return;

    const dateRead = currentQuote.book.date_read;
    const hintDisplay = document.getElementById("hintDisplay");

    // Parse the date (format: "MMM DD, YYYY" like "Feb 25, 2014")
    const date = new Date(dateRead);
    const year = date.getFullYear();

    // Create a 2-year range around the read date
    const startYear = year - 1;
    const endYear = year + 1;

    hintDisplay.textContent =
        `📅 You read this book between ${startYear} and ${endYear}`;
    hintDisplay.style.display = "block";

    hintsUsed.push("time");
    this.disabled = true;
});

function fuzzyMatch(query, text) {
    query = query.toLowerCase();
    text = text.toLowerCase();

    // Check if text starts with query
    if (text.startsWith(query)) {
        return 1.0;
    }

    // Check if any word in text starts with query
    const words = text.split(/\s+/);
    for (let word of words) {
        if (word.startsWith(query)) {
            return 0.9;
        }
    }

    // Check if query is contained in text
    if (text.includes(query)) {
        return 0.8;
    }

    // Fall back to Levenshtein distance
    const cleanQuery = query.replace(/[^a-z0-9]/g, "");
    const cleanText = text.replace(/[^a-z0-9]/g, "");

    const matrix = [];
    for (let i = 0; i <= cleanText.length; i++) matrix[i] = [i];
    for (let j = 0; j <= cleanQuery.length; j++) matrix[0][j] = j;

    for (let i = 1; i <= cleanText.length; i++) {
        for (let j = 1; j <= cleanQuery.length; j++) {
            matrix[i][j] = cleanText[i - 1] === cleanQuery[j - 1]
                ? matrix[i - 1][j - 1]
                : Math.min(
                    matrix[i - 1][j - 1] + 1,
                    matrix[i][j - 1] + 1,
                    matrix[i - 1][j] + 1,
                );
        }
    }

    const distance = matrix[cleanText.length][cleanQuery.length];
    return Math.max(
        0,
        1 - distance / Math.max(cleanQuery.length, cleanText.length),
    );
}

function showSuggestions(inputId, listId, key) {
    const input = document.getElementById(inputId);
    const list = document.getElementById(listId);
    let currentMatches = [];
    let selectedIndex = 0;

    const selectMatch = (match) => {
        input.value = match.text;
        list.style.display = "none";
        selectedIndex = 0;

        // Auto-populate author if title was selected
        if (key === "title") {
            document.getElementById("author").value = match.book.author;
        }
    };

    const updateSelection = () => {
        const items = list.querySelectorAll("li");
        items.forEach((item, index) => {
            if (index === selectedIndex) {
                item.classList.add("selected");
            } else {
                item.classList.remove("selected");
            }
        });
    };

    const updateSuggestions = () => {
        const query = input.value.trim().toLowerCase();
        list.innerHTML = "";
        selectedIndex = 0;
        if (!query) {
            list.style.display = "none";
            currentMatches = [];
            return;
        }

        // Fuzzy match
        const matches = books
            .map((book) => ({
                text: book[key],
                book: book,
                score: fuzzyMatch(query, book[key]),
            }))
            .filter((item) => item.score > 0.3)
            .sort((a, b) => b.score - a.score)
            .slice(0, 8);

        currentMatches = matches;

        if (matches.length === 0) {
            list.style.display = "none";
            return;
        }

        matches.forEach((match, index) => {
            const li = document.createElement("li");
            li.textContent = match.text;
            if (index === 0) li.classList.add("selected");
            li.addEventListener("click", () => selectMatch(match));
            list.appendChild(li);
        });

        list.style.display = "block";
    };

    input.addEventListener("input", updateSuggestions);

    // Handle keyboard navigation
    input.addEventListener("keydown", (e) => {
        if (currentMatches.length === 0) return;

        if (e.key === "ArrowDown") {
            e.preventDefault();
            selectedIndex = (selectedIndex + 1) %
                currentMatches.length;
            updateSelection();
        } else if (e.key === "ArrowUp") {
            e.preventDefault();
            selectedIndex = (selectedIndex - 1 + currentMatches.length) %
                currentMatches.length;
            updateSelection();
        } else if (e.key === "Enter" || e.key === "Tab") {
            if (list.style.display === "block") {
                e.preventDefault();
                selectMatch(currentMatches[selectedIndex]);
            }
        } else if (e.key === "Escape") {
            list.style.display = "none";
            selectedIndex = 0;
        }
    });

    // Hide when clicking away
    document.addEventListener("click", (e) => {
        if (!list.contains(e.target) && e.target !== input) {
            list.style.display = "none";
            selectedIndex = 0;
        }
    });

    // Return function to trigger suggestions externally
    return { updateSuggestions };
}

// Store references to update functions
const titleSuggestions = showSuggestions("title", "titleSuggestions", "title");
const authorSuggestions = showSuggestions(
    "author",
    "authorSuggestions",
    "author",
);

// Handle author selection
document.getElementById("author").addEventListener("change", () => {
    const selectedAuthor = document.getElementById("author").value.trim();
    if (!selectedAuthor) return;

    // Find all books by this author
    const booksByAuthor = books.filter((book) =>
        book.author.toLowerCase() === selectedAuthor.toLowerCase()
    );

    const titleInput = document.getElementById("title");

    if (booksByAuthor.length === 1) {
        // Auto-populate if only one book
        titleInput.value = booksByAuthor[0].title;
    } else if (booksByAuthor.length > 1) {
        // Show suggestions for this author's books
        titleInput.focus();
        titleInput.dispatchEvent(new Event("input"));
    }
});

// Load data on page load
loadData();
